pkg/upload/js.go

```diff
@@ -16,6 +24,158 @@ type JsOptions struct {
 	Path        utils.Paths `arg:"" name:"path" help:"Path to a directory of source maps and bundles to upload" type:"path" default:"."`
 }
 
+// Resolve the project root by walking up from the path specified until a package.json is found
+func resolveProjectRoot(jsOptions JsOptions, path string) (string, error) {
+	if jsOptions.ProjectRoot != "" {
+		return jsOptions.ProjectRoot, nil
+	}
+	checkPath, err := filepath.Abs(path)
+	if err != nil {
+		return "", fmt.Errorf("unable to make project root an absolute path %s", path)
+	}
+	// Walk up the folder structure as far as possible
+	for filepath.Dir(checkPath) != checkPath {
+		packageJson := filepath.Join(checkPath, "package.json")
+		if !utils.FileExists(packageJson) {
+			checkPath = filepath.Dir(checkPath)
+			continue
+		}
+		return checkPath, nil
+	}
+	return path, nil
+}
+
+// Attempt to parse information from the package.json file if values aren't provided on the command line
+func ResolveVersion(jsOptions JsOptions, projectRoot string, logger log.Logger) (string, error) {
+	if jsOptions.VersionName != "" {
+		return jsOptions.VersionName, nil
+	}
+	checkPath, err := filepath.Abs(projectRoot)
+	if err != nil {
+		return "", fmt.Errorf("unable to make project root an absolute path %s", projectRoot)
+	}
+	// Walk up the folder structure as far as possible
+	for filepath.Dir(checkPath) != checkPath {
+		packageJson := filepath.Join(checkPath, "package.json")
+		if !utils.FileExists(packageJson) {
+			checkPath = filepath.Dir(checkPath)
+			continue
+		}
+		file, err := os.Open(packageJson)
+		if err != nil {
+			return "", fmt.Errorf("unable to open package.json %s", packageJson)
```

Comment:
The approach we've taken for other commands is to _attempt_ to find data from the environment but don't present an error in doing so as parsing the `package.json` is a secondary activity that the user might not be expected – print Info/Warning messages if information is not found in the file, but we shouldn't error out at this point as an error saying "unable to parse package.json" won't make much sense to the user.

So we put the validation for required data at the very last moment, just before the upload, and only then present an error to say "no version found".

Reply:
Sounds reasonable. The API doesn't need a version anyway so we can just skip the version if it is not specified by the user or found.